cranelift/codegen/src/isa/x64/inst/emit.rs

```diff
@@ -4089,3 +4139,156 @@ pub(crate) fn emit(
+    // onto our current frame, using only volatile, non-argument
+    // registers.
+    //
+    //
+    // The current stack layout is the following:
+    //
+    //            | ...                 |
+    //            +---------------------+
+    //            | ...                 |
+    //            | stack arguments     |
+    //            | ...                 |
+    //    current | return address      |
+    //    frame   | old FP              | <-- FP
+    //            | ...                 |
+    //            | old stack slots     |
+    //            | ...                 |
+    //            +---------------------+
+    //            | ...                 |
+    //    new     | new stack arguments |
+    //    frame   | ...                 | <-- SP
+    //            +---------------------+
+    //
+    // We need to pop FP, copy the new stack arguments over the old
+    // stack arguments, write the return address into the correct slot
+    // just after the new stack arguments, adjust SP to point to the new
+    // return address, and then jump to the callee (which will push the
+    // old FP again).
+
+    // "Pop" the old FP into `rbp`.
+    Inst::Mov64MR {
+        src: SyntheticAmode::Real(Amode::ImmReg {
+            simm32: 0,
+            base: fp,
+            flags: MemFlags::trusted(),
+        }),
+        dst: Writable::from_reg(Gpr::new(regs::rbp()).unwrap()),
+    }
+    .emit(&[], sink, info, state);
+
+    // The new lowest address (top of stack) -- relative to FP -- for
+    // our tail callee. We compute this now so that we can move our
+    // stack arguments into place.
+    let callee_sp_relative_to_fp = old_stack_arg_size.wrapping_sub(new_stack_arg_size);
+
+    // Copy over each word, using `tmp` as a temporary register.
+    //
+    // Note that we have to do this from stack slots with the highest
+    // address to lowest address because in the case of when the tail
+    // callee has more stack arguments than we do, we might otherwise
+    // overwrite some of our stack arguments before they've been copied
+    // into place.
+    assert_eq!(
+        new_stack_arg_size % 8,
+        0,
+        "stack argument space sizes should always be 8-byte aligned"
+    );
+    for i in (0..new_stack_arg_size / 8).rev() {
+        Inst::Mov64MR {
+            src: SyntheticAmode::Real(Amode::ImmReg {
+                simm32: i * 8,
+                base: regs::rsp(),
+                flags: MemFlags::trusted(),
+            }),
+            dst: tmp_w,
+        }
+        .emit(&[], sink, info, state);
+        Inst::MovRM {
+            size: OperandSize::Size64,
+            src: tmp,
+            dst: SyntheticAmode::Real(Amode::ImmReg {
+                // Add 2 because we need to skip over the old FP and the
+                // return address.
+                simm32: callee_sp_relative_to_fp.wrapping_add((i + 2) * 8),
+                base: fp,
```

Comment:
I'm possibly confused here, but: isn't this using the wrong FP, i.e., the value we already restored above, to whatever the original caller had in RBP when we were called? I think we want to use *our* FP as a frame of reference here.

Reply:
Ah, actually, after reading further, I see that this is `fp` which is set up in the lowering helper to be a copy (`MovFromPReg`) of `rbp`, so this makes sense.

Reply:
(Oh I just hadn't refreshed the tab. Doh.)